Introduce the cfb_shell sample app

diegosueiro opened the pull request:
This PR adds:

- `cfb_framebuffer_invert` function to invert the pixels in the character framebuffer;
- `cfb_get_numof_fonts` function to get the number of available fonts 
- a character framebuffer shell module;
- a sample app to test the character framebuffer shell module.

All tested and validated with the reel_board: `cmake -DBOARD=reel_board  ../..`

diegosueiro committed:
subsys/fb/cfb: Introduce cfb_framebuffer_invert function

Adds the cfb_framebuffer_invert function to invert Pixels in the
Character Framebuffer

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

vanwinkeljan commented on subsys/fb/Kconfig:
```diff
@@ -21,6 +21,14 @@ config CHARACTER_FRAMEBUFFER_USE_DEFAULT_FONTS
 
+config CHARACTER_FRAMEBUFFER_SHELL
+	bool "Character Framebuffer shell"
+	select SERIAL
```
No need to depend on serial, the correct backend needs to be configured in the shell Kconfig

vanwinkeljan commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,411 @@
+#define HELP_INIT "call \"cfb init\" first"
+
+#if defined(CONFIG_SSD1673)
+#define DISPLAY_DRIVER				"SSD1673"
```
This should come from Kconfig (e.g. CHARACTER_FRAMEBUFFER_SHELL_DISPLAY_NAME) now it limits the shell to only two drivers and further it assumes the names are hard coded to ```SSD1673``` and ```SSD1306```

vanwinkeljan commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,411 @@
+#endif
+
+
+const struct shell *ctx_shell;
```
What is the use of this global variable? As far as I can see it is only used in the macros ```print``` and ```error``` but then these macros are always passed a shell pointer.

vanwinkeljan commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,411 @@
+
+
+const struct shell *ctx_shell;
+struct device *dev;
```
Should be static

vanwinkeljan commented on subsys/fb/cfb_shell.c:
```diff
@@ -0,0 +1,411 @@
+		return -ENODEV;
+	}
+
+	for (int idx = 0; idx < 42; idx++) {
```
Where does this 42 comes from, could this not be a define?

vanwinkeljan reviewed (changes requested):
On top of the inline comments please add a README.rst for the sample.

diegosueiro committed:
subsys/fb/cfb: Introduce cfb_get_numof_fonts function

Adds the cfb_get_numof_fonts function to get the number of
available fonts

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

vanwinkeljan commented:
@diegosueiro Code looks good to me but could you add a README.rst file to the sample then it will be visible from zephyrs documentation?

diegosueiro commented:
@vanwinkeljan and @jfischer-phytec-iot ,

Any more comments on this PR?

vanwinkeljan commented:
@diegosueiro No only open comment from my side is the missing README.rst

nashif commented on samples/display/cfb_shell/prj.conf:
```diff
@@ -0,0 +1,14 @@
+CONFIG_DISPLAY=y
+
+CONFIG_LOG=y
+CONFIG_DISPLAY_LOG_LEVEL_DBG=y
```
do not enable debug logging by default.

nashif commented on samples/display/cfb_shell/prj.conf:
```diff
@@ -0,0 +1,14 @@
+CONFIG_LOG=y
+CONFIG_DISPLAY_LOG_LEVEL_DBG=y
+
+CONFIG_CFB_LOG_LEVEL_DBG=y
```
ditto

diegosueiro commented:
@nashif ,

Do you think we should move from `samples/display/cfb_shell` to `samples/subsys/shell/cfb_module/` ?

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,47 @@
+
+Overview
+********
+A simple shell module that exercises displays using the Character Framebuffer
```
```suggestion
This is a simple shell module that exercises displays using the Character Framebuffer
```

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,47 @@
+Building and Running
+********************
+
+Build the sample app by choosing the OVERLAY_CONFIG for the target board. E.g.:
```
```suggestion
Build the sample app by choosing the OVERLAY_CONFIG for the target board, for example:
```

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,47 @@
+Shell Module Command Help
+=========================
+
+.. code-block:: console
```
While this shows the command syntax, it doesn't explain what the (sub)commands actually do.

diegosueiro replied:
Aren't the command names self-explanatory?

nashif replied:
I agree with @dbkinder here, anything is better than "[none]" and even for those with help, it is not really easy for someone not familiar with the feature to do anything simple.

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,46 @@
+                 -h, --help  :Show command help.
+         Subcommands:
+                 init        :[none]
+                 get_param   :<all, heigh, width, ppt, rows, cols>
```
``height``

diegosueiro commented:
@dbkinder and @nashif ,

Any more comments?

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,76 @@
+
+**get_device**: prints the display device name.
+
+**get_param**: get the display parameters where heigh, width and ppt
```
height

dbkinder reviewed (dismissed):
LGTM but one typo

diegosueiro committed:
shell: Add initial code for Character Framebuffer shell module

This adds the config option and files of the Character Framebuffer
shell module

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro committed:
samples/display: Introduce the cfb_shell sample app

This adds the shell sample app for the Character Framebuffer testing.

cfb - Character Framebuffer shell commands
Options:
  -h, --help  :Show command help.
Subcommands:
  init        :[none]
  get_device  :[none]
  get_param   :<all, height, width, ppt, rows, cols>
  get_fonts   :[none]
  set_font    :<idx>
  invert      :[none]
  print       :<col: pos> <row: pos> <text>
  scroll      :<dir: (vertical|horizontal)> <col: pos> <row: pos> <text>
  clear       :[none]

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

nashif commented:
@diegosueiro Can you provide as an example a list of commands that can entered on the shell to allow someone unfamiliar to get going quickly, like print "reel board" in different way, scroll it, invert it, flip it, you name it....

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,75 @@
+
+**print**: pass the initial column and row positions and the text in
+double quotation marks when it contains spaces. If text hits the edge
+the display the remain characters will be displayed in the next line. The
```
```suggestion
of the display the remaining characters will be displayed on the next line. The
```

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,75 @@
+(pixel per tile) are in pixels and the number of rows and columns. The row
+position is incremented by a multiple of the ppt.
+
+**get_fonts**: print the index, height and width in pixels of the static
```
how do you install a new font?

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,75 @@
+display boundary, last column for horizontal and last row for vertical
+direction.
+
+**clear**: clear the display screen.
```
It would be great, after these descriptions, to show a nice example of how to display a phrase on the display and manipulate it (and have pictures or describe what gets displayed).

diegosueiro replied:
I can put command examples as @nashif suggested, but at the moment there is no API, AFAIK, to draw pictures, or manipulated what is printed like inversion.

vanwinkeljan replied:
@diegosueiro I think what @dbkinder means is to give examples with screenshots in the docs.
But on adding the screenshots in the docs I already raised the question (PR #9149) if it is allowed to add binary files into the git repo. In my opinion this is not a good idea and git-lfs is not setup for the zephyr repo (@nashif any inputs on this?)

Regarding the API for drawing you have PR #6826 if you need something more fancy than CFB. CFB is just a lightweight framework for displaying text where LVGL is a full blown GUI library and I think they complement each other nicely

dbkinder reviewed (changes requested):
Sorry to pick on your sample, but there's no documentation elsewhere about the CFB API.  Maybe some of this wouldn't be necessary if that were documented.

diegosueiro commented:
> @diegosueiro Can you provide as an example a list of commands that can entered on the shell to allow someone unfamiliar to get going quickly, like print "reel board" in different way, scroll it, invert it, flip it, you name it....

I will add examples.

Do you think we should move from `samples/display/cfb_shell/` to `samples/subsys/shell/cfb_module/`

diegosueiro commented:
@dbkinder ,

Anything more is needed?

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,142 @@
+
+         uart:~$ cfb print 60 5 ZEPHYR
+
+**scroll**: pass the scroll direction, vertical or horizontal, the initial
```
Thanks for the update and examples!  I'm still a bit confused about the difference between "print" and "scroll".  Is "print" the same as "scroll vertical"?

diegosueiro replied:
@dbkinder,
 The scroll command will "move" vertically or horizontally the text from the initial position to the display boundaries. And the print will only display the text without any movement.

dbkinder replied:
Ah, so "scroll" inserts the text and moves any existing text on the display, and "print" will overwrite existing text?  That's worth clarifying.

diegosueiro replied:
Nope. The scroll will move the text that you passed with the scroll command. What is already on the display will be overwritten as well.

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,142 @@
+**print**: pass the initial column and row positions and the text in
+double quotation marks when it contains spaces. If text hits the edge
+of the display the remaining characters will be displayed on the next line. The
+previous printed text will be erased.
```
```suggestion
previous printed text will be overwritten.
```

dbkinder commented on samples/display/cfb_shell/README.rst:
```diff
@@ -0,0 +1,142 @@
+contains spaces. If text hits the edge the display the remaining characters
+will be displayed in the next line. The text will scroll until it hits the
+display boundary, last column for horizontal and last row for vertical
+direction.
```
```suggestion
direction.  Existing text on the display will be moved, potentially off the display.
```

dbkinder reviewed (dismissed):
Did I get that right?

diegosueiro commented:
@dbkinder,

Any more comments on the documentation?
